handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1722,6 +1730,69 @@ private ByteBuf allocateOutNetBuf(ChannelHandlerContext ctx, int pendingBytes, i
         return allocate(ctx, engineType.calculateWrapBufferCapacity(this, pendingBytes, numComponents));
     }
 
+    static boolean attemptCopyToCumulation(ByteBuf cumulation, ByteBuf next) {
+        final int inReadableBytes = next.readableBytes();
+        final int cumCapacity = cumulation.capacity();
+        if (MAX_PLAINTEXT_LENGTH - cumulation.readableBytes() >= inReadableBytes &&
+                (cumulation.isWritable(inReadableBytes) && cumCapacity >= MAX_PLAINTEXT_LENGTH ||
+                        cumCapacity < MAX_PLAINTEXT_LENGTH &&
+                                ensureWritableSuccess(cumulation.ensureWritable(inReadableBytes, false)))) {
+            cumulation.writeBytes(next);
+            next.release();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Each call to SSL_write will introduce about ~100 bytes of overhead. This coalescing queue attempts to increase
+     * goodput by aggregating the plaintext in chunks of {@link #MAX_PLAINTEXT_LENGTH}. If many small chunks are written
+     * this can increase goodput, decrease the amount of calls to SSL_write, and decrease overall encryption operations.
+     */
+    private static final class SslHandlerCoalescingBufferQueue extends AbstractCoalescingBufferQueue {
+        SslHandlerCoalescingBufferQueue(int initSize) {
+            super(initSize);
+        }
+
+        @Override
+        protected ByteBuf compose(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf next) {
+            if (cumulation instanceof CompositeByteBuf) {
+                CompositeByteBuf composite = (CompositeByteBuf) cumulation;
+                int numComponents = composite.numComponents();
+                if (numComponents == 0 ||
+                        !attemptCopyToCumulation(composite.internalComponent(numComponents - 1), next)) {
+                    composite.addComponent(true, next);
+                }
+                return composite;
+            }
+            if (attemptCopyToCumulation(cumulation, next)) {
+                return cumulation;
+            }
+            CompositeByteBuf composite = alloc.compositeDirectBuffer(size() + 2);
+            composite.addComponent(true, cumulation);
+            composite.addComponent(true, next);
+            return composite;
+        }
+
+        @Override
+        protected ByteBuf composeFirst(ByteBufAllocator allocator, ByteBuf first) {
+            // TODO(scott): should this be configurable? for example maybe only flatten if some flag is set, the
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)